app/serializers/api/admin/order_cycle_serializer.rb

```diff
@@ -74,4 +61,28 @@ def visible_variants_for_outgoing_exchanges
     end
     visible
   end
+
+  private
+
+  def permissions
+    @permissions ||= OpenFoodNetwork::OrderCyclePermissions.new(options[:current_user], object)
+  end
+
+  def visible_enterprises
+    @visible_enterprises ||= permissions.visible_enterprises
+  end
+
+  def sort_by_supplier_id(variants)
+    collection = {}
+    variants.map do |variant|
+      supplier_id = variant.product.supplier_id
+
+      if collection.key? supplier_id
+        collection[supplier_id] << variant.id
+      else
+        collection[supplier_id] = [variant.id]
+      end
+    end
+    collection
+  end
```

Comment:
we got to start trusting our DB to work with data if we ever want to have decent performance. I know it's sometimes hard but there's no way around it. Have you tried with something like

```rb
variants
  .joins(:product)
  .order(:supplier_id)
  .select('supplier_id, spree_variants.*')

variants.group_by(&:supplier_id)
```

It'll return a hash like

```rb
"205"=>
  [#<Spree::Variant id: 3015, sku: "", weight: nil, height: nil, width: nil, depth: nil, deleted_at: nil, is_master: true, product_id: 1391, cost_price: nil, position: 1, cost_currency: "EUR", unit_value: 1.0, unit_description: "", display_name: nil, display_as: nil, import_date: nil>,                                                                               
   #<Spree::Variant id: 3016, sku: "", weight: nil, height: nil, width: nil, depth: nil, deleted_at: nil, is_master: false, product_id: 1391, cost_price: nil, position: 2, cost_currency: "EUR", unit_value: 1.0, unit_description: "", display_name: nil, display_as: nil, import_date: nil>]}
```

Because I assume that's the structure we need for the client, right? `Enumerable#group_by` does have a cost...

Reply:
The hash has to be variant ids grouped by supplier ids, but yes I'll take a look at tidying it up.

Reply:
I think that we should change the name because it's not sorting anything. It's `group_variant_ids_by_supplier_id`.

I like Pau's recommendation to use group_by. Changing the variant objects to their ids makes it a bit more complicated. But if it needs to be:

```rb
variants.group_by(&:supplier_id).map { |supplier_id, variants| [supllier_id, variants.map(&:id)] }.to_h
```

It's still a bit complicated and apparently using a lot of memory. A way to use less would be this:
```rb
grouped = variants.group_by(&:supplier_id)
grouped.each { |supplier_id, variants|  grouped[supllier_id] = variants.map(&:id) }

# or shorter but lesser known method:
grouped.update(grouped) { |_, variants| variants.map(&:id) }
```

This is a lot simpler and then grouping is probably very fast in Ruby. But it's difficult to tell if it's actually better than Matt's version which iterates only once and doesn't create any intermediate objects.

I'm okay with keeping the current version, just rename the method. And in both cases we can do a little optimisation:

```rb
variants.select("spree_variants.id, spree_products.supplier_id")
```

But that's only good if we don't re-use the same relation somewhere else. Otherwise we will trigger an additional query here.